theano/compile/function_module.py

```diff
@@ -488,16 +489,195 @@ def __setitem__(self, item, value):
+            i.variable = in_var
+            if not delete_updates and i.update is not None:
+                i.update = fg_cpy.outputs[update_i]
+                update_i += 1
+            else:
+                i.update = None
+
+        # Construct new storage_map that map new variable to old storage,
+        # so that the ensuing function shares storage with the original one
+        storage_map = self.fn.storage_map
+        new_storage_map = {}
+        # TODO: We could share the output storage, but we must make sure
+        # 2 different function call won't override each other values. This
+        # is already done elsewhere, so to reuse it the user would need to
+        # use Out(var, borrow=True) and maybe the mutable=True flag too.
+        # But to be safe for now as it isn't documented and we aren't sure
+        # it is well tested, we don't share the part of the storage_map.
+        if share_memory:
+            i_o_vars = maker.fgraph.inputs + maker.fgraph.outputs
+            for key in storage_map.keys():
+                if key not in i_o_vars:
+                    new_storage_map[memo[key]] = storage_map[key]
+
+        if not name and self.name:
+            name = self.name + " copy"
+
+        input_storage = [i.value for i in ins]
+        # reinitialize new maker and create new function
+        if profile is None:
+            profile = config.profile
+            # profile -> True or False
+        if profile is True:
+            if name:
+                message = name
             else:
-                there.data = here.data
-        return cpy
+                message = str(maker.profile.message) + " copy"
+            profile = theano.compile.profiling.ProfileStats(message=message)
+            # profile -> object
+        elif type(profile) == str:
+            profile = theano.compile.profiling.ProfileStats(message=profile)
+
+        f_cpy = maker.__class__(inputs=ins, outputs=outs, fgraph=fg_cpy,
+                                mode=maker.mode, profile=profile,
+                                on_unused_input=maker.on_unused_input,
+                                function_builder=maker.function_builder,
+                                accept_inplace=maker.accept_inplace
+                                ).create(input_storage,
+                                         storage_map=new_storage_map)
+
+        for in_ori, in_cpy, ori, cpy in zip(maker.inputs, f_cpy.maker.inputs,
+                                            self.input_storage,
+                                            f_cpy.input_storage):
+
+            # Share immutable ShareVariable and constant input's storage
+            swapped = swap is not None and in_ori.variable in swap_svs_ori
+
+            # Using the original storage if SharedVariable will not be updated
+            # and is not swapped
+            if not in_ori.mutable and not swapped:
+                cpy.data = ori.data
+                in_cpy.value = in_ori.value
```

Comment:
condition ```not swapped``` is to make sure not to share container of swapped ```SharedVariable```. I do not copy the container of any ```SharedVariable``` so I think here it's ok to not enter the block for not swapped ```SharedVariable```.

--------------------

condition ```not in_ori.mutable``` is to make sure not to copy container of ```In``` instances. This is to support the original test ```test_copy()``` where the function was defined using ```In``` rather than variables.